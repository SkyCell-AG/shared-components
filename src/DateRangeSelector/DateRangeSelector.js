import React from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'

import DateSelect from 'DateSelect'
import rangeToDayPassed from 'utils/rangeToDayPassed'
import {
    strToDate,
} from 'utils/DateUtils'

import useStyles from './DateRangeSelector.style'

const propTypes = {
    mini: PropTypes.bool,
    options: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string.isRequired,
        value: PropTypes.number.isRequired,
    })),
    selectOption: PropTypes.func.isRequired,
    onChangeRange: PropTypes.func.isRequired,
    showTimeRange: PropTypes.bool,
    from: PropTypes.string.isRequired,
    to: PropTypes.string.isRequired,
}

const defaultProps = {
    mini: false,
    showTimeRange: false,
    options: [
        {
            label: '24h',
            value: 1,
        },
        {
            label: '7d',
            value: 7,
        },
        {
            label: '14d',
            value: 14,
        },
    ],
}

const DateRangeSelector = ({
    mini,
    options,
    selectOption,
    showTimeRange,
    onChangeRange,
    from,
    to,
}) => {
    const classes = useStyles()
    const dayPassed = from && to ? rangeToDayPassed({
        from: strToDate(from),
        to: strToDate(to),
    }) : undefined

    return (
        <div className={classes.wrapper}>
            {!mini && (
                <div className={classes.optionsWrapper}>
                    {options.map(({
                        label,
                        value,
                    }) => {
                        return (
                            <div
                                data-testid={`date-range-button-${value}`}
                                key={value}
                                className={clsx(classes.option, {
                                    [classes.selectedOption]: dayPassed === value,
                                })}
                                onMouseDown={() => {
                                    selectOption(value)
                                }}
                            >
                                {label}
                            </div>
                        )
                    })}
                </div>
            )}
            {showTimeRange && (
                <div className={classes.inputsWrapper}>
                    <div className={classes.input}>
                        <DateSelect
                            title="From"
                            className={classes.datePicker}
                            onChange={onChangeRange}
                            name="from"
                            value={from}

                        />
                    </div>
                    <div className={classes.input}>
                        <DateSelect
                            title="To"
                            className={classes.datePicker}
                            onChange={onChangeRange}
                            name="to"
                            value={to}
                        />
                    </div>
                </div>
            )}
        </div>
    )
}

DateRangeSelector.propTypes = propTypes
DateRangeSelector.defaultProps = defaultProps

export default DateRangeSelector
