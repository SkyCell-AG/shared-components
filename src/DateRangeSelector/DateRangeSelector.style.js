import {
    makeStyles,
} from '@mui/styles'
import {
    useTheme,
} from '@mui/material/styles'

const useStyles = makeStyles(() => {
    const theme = useTheme()

    return {
        wrapper: {
            display: 'flex',
            justifyContent: 'space-between',
            fontSize: theme.typography.h4.fontSize,
        },

        optionsWrapper: {
            display: 'flex',
            alignItems: 'center',
        },

        option: {
            cursor: 'pointer',
            padding: theme.spacing(0.625),
            width: 40,
            textAlign: 'center',
            transition: 'background-color 03:s',
            '&:hover': {
                backgroundColor: theme.palette.secondary[100],
            },
        },

        selectedOption: {
            backgroundColor: `${theme.palette.secondary[100]} !important`,
        },

        inputsWrapper: {
            display: 'flex',
        },

        input: {
            alignItems: 'center',
            marginLeft: theme.spacing(2.5),
        },

        dateContainer: {
            display: 'flex',
            fontSize: '1.2rem',
        },

        editedDate: {
            color: 'silver',
        },

        dayPickerContainer: {
            display: 'none',
            position: 'absolute',
            width: '550px',
            marginLeft: '-310px',
        },

        dayPickerOpened: {
            display: 'block',
        },

        datePicker: {
            marginLeft: theme.spacing(1.25),
        },
    }
})

export default useStyles
