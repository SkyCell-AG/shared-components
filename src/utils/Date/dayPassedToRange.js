import moment from 'moment'

function dayPassedToRange(value) {
    const to = new Date()

    const from = moment().subtract(value, 'days').toDate()

    return {
        from,
        to,
    }
}

export default dayPassedToRange
