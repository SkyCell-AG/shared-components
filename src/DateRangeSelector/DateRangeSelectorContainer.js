import React, {
    useCallback,
} from 'react'
import PropTypes from 'prop-types'
import noop from 'lodash/noop'

import {
    dayPassedToRange,
    dateToDateStr,
} from 'utils/DateUtils'

import DateRangeSelector from './DateRangeSelector'

const propTypes = {
    value: PropTypes.shape({
        from: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        to: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
    }).isRequired,
    onChange: PropTypes.func,
    setDateRange: PropTypes.func.isRequired,
}

const defaultProps = {
    onChange: noop,
}

const DateRangeSelectorContainer = ({
    value = {},
    onChange,
    setDateRange,
    ...rest
}) => {
    const {
        from,
        to,
    } = value

    const selectOption = useCallback((selected) => {
        onChange(dayPassedToRange(selected))
        setDateRange(false)
    }, [
        onChange,
        setDateRange,
    ])

    const onChangeRange = useCallback((_, {
        target: {
            name,
            value: newValue,
        },
    }) => {
        const updatedData = {
            from: new Date(from).setHours(0, 0, 0),
            to: new Date(to).setHours(23, 59, 59),
            [name]: newValue,
        }

        onChange(updatedData, {
            name,
            value: updatedData,
        })

        setDateRange(true)
    }, [
        from,
        onChange,
        setDateRange,
        to,
    ])

    return (
        <DateRangeSelector
            {...rest}
            selectOption={selectOption}
            onChangeRange={onChangeRange}
            from={dateToDateStr(from)}
            to={dateToDateStr(to)}
        />
    )
}

DateRangeSelectorContainer.propTypes = propTypes
DateRangeSelectorContainer.defaultProps = defaultProps

export default DateRangeSelectorContainer
