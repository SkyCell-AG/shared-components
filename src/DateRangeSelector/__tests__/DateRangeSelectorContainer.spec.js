import React from 'react'
import {
    shallow,
} from 'enzyme'

import dayPassedToRange from 'utils/DateUtils/dayPassedToRange'

import DateRangeSelectorContainer from '../DateRangeSelectorContainer'
import DateRangeSelector from '../DateRangeSelector'

jest.mock('utils/DateUtils/dayPassedToRange', () => {
    return jest.fn()
})

describe('shared-components/DateRangeSelectorContainer', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(
            <DateRangeSelectorContainer
                setDateRange={jest.fn()}
                value={{
                    from: new Date(2020, 8, 16).getTime(),
                    to: new Date(2020, 8, 18).getTime(),
                }}
                onChange={jest.fn()}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('onChange', () => {
        const onChange = jest.fn()
        const setRange = jest.fn()

        const initValue = {
            from: new Date(2020, 8, 16).setHours(0, 0, 0),
            to: new Date(2020, 8, 18).setHours(23, 59, 59),
        }

        const wrapper = shallow(<DateRangeSelectorContainer
            setDateRange={setRange}
            onChange={onChange}
            value={initValue}
        />)
            .find(DateRangeSelector)

        const newFrom = new Date(2020, 8, 19).setHours(0, 0, 0)

        wrapper.props().onChangeRange(null, {
            target: {
                name: 'from',
                value: newFrom,
            },
        })

        const newValue = {
            ...initValue,
            from: newFrom,
        }

        expect(onChange).toBeCalledWith(newValue, {
            name: 'from',
            value: {
                from: newFrom,
                to: initValue.to,
            },
        })
        expect(setRange).toBeCalledWith(true)
    })

    it('on select option', () => {
        dayPassedToRange.mockReturnValue({
            from: 'from',
            to: 'to',
        })
        const onChange = jest.fn()
        const setRange = jest.fn()

        const initValue = {
            from: new Date(2020, 8, 16).setHours(0, 0, 0),
            to: new Date(2020, 8, 18).setHours(23, 59, 59),
        }

        const wrapper = shallow(<DateRangeSelectorContainer
            setDateRange={setRange}
            onChange={onChange}
            value={initValue}
        />)
            .find(DateRangeSelector)

        wrapper.props().selectOption('24h')

        expect(dayPassedToRange).toBeCalledWith('24h')
        expect(onChange).toBeCalledWith({
            from: 'from',
            to: 'to',
        })
        expect(setRange).toBeCalledWith(false)
    })
})
