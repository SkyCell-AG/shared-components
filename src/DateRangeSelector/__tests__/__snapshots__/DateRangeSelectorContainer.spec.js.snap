// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared-components/DateRangeSelectorContainer should match snapshot 1`] = `
<DateRangeSelector
  from="16.09.2020"
  mini={false}
  onChangeRange={[Function]}
  options={
    Array [
      Object {
        "label": "24h",
        "value": 1,
      },
      Object {
        "label": "7d",
        "value": 7,
      },
      Object {
        "label": "14d",
        "value": 14,
      },
    ]
  }
  selectOption={[Function]}
  showTimeRange={false}
  to="18.09.2020"
/>
`;
